### PR TITLE
Setting replyto in the header

### PR DESCRIPTION
**Pull request by tushdante**
- If name is available, set replyto in the header
- If not, set in the WebAPI
- Tests still failing, not sure why.

### CODE REVIEW

**Comment by tushdante** [comment]
TODO: Version bump; Update Readme

File: sendgrid/message.py [hunk 1]
```diff
@@ -146,10 +163,19 @@ def add_content_id(self, cid, value):
         self.content[cid] = value
 
     def set_headers(self, headers):
+        # if not headers:
```
Comment (tgwizard) [inline sendgrid/message.py]: Perhaps these comments should be removed?

File: sendgrid/message.py [hunk 2]
```diff
@@ -123,7 +125,22 @@ def add_bcc(self, bcc):
...
-        self.reply_to = replyto
+        name, email = rfc822.parseaddr(replyto.replace(',', ''))
+        if name and email:
+            self.set_reply_to_name(replyto)
+        elif email:
+            self.reply_to = email
+
+    def set_reply_to_name(self, replyto):
+        if sys.version_info < (3, 0) and isinstance(replyto, unicode):
+            replyto = replyto.encode('utf-8')
```
Comment (tgwizard) [inline sendgrid/message.py]: Should this encoding be done for all header values? If so, would it make sense to move this logic to the `set_headers` method?

File: sendgrid/message.py [hunk 3]
```diff
@@ -146,10 +163,19 @@ def add_content_id(self, cid, value):
...
         if isinstance(headers, str):
-            self.headers = headers
-        else:
-            self.headers = json.dumps(headers)
+            headers = json.loads(headers)
+        for key, value in headers.iteritems():
+            self.headers[key] = value
+
+        # Convert back to string
+        self.headers = json.dumps(self.headers)
```
Comment (tgwizard) [inline sendgrid/message.py]: Isn't it simpler to keep `self.headers` as a dict while in memory, and then converting to string via `json.dumps` when serializing to send to the SendGrid API? Or will this break backwards compatibility?

**Comment by tgwizard** [comment]
The tests fail now because previously, when you set no headers, you had `self.headers = ''`. Now `self.headers` will default to `'{}'` (the result of `json.dumps('{}')`). In `SendGridClient._build_body`, you remove all Falsy fields before sending them: https://github.com/sendgrid/sendgrid-python/blob/33039c102d6ba6ca5869f1f743a4888a3fbbaa76/sendgrid/sendgrid.py#L84

```python
>>> bool('')
False
>>> bool('{}')
True
```

**Comment by tushdante** [comment]
Thanks @tgwizard ! I appreciate the feedback. I'll make the changes and try again

File: sendgrid/message.py (hunk 3 above)
Comment (tushdante) [inline sendgrid/message.py]: @tgwizard The reason I've had to to do the conversion is so as to not break backwards compatibility.

**Comment by tushdante** [comment]
Also, @tgwizard wouldn't serializing the headers to a string (even with an empty dict) make it Truthy?

**Comment by tushdante** [comment]
Hey @thinkingserious looks like the issue is with the mock code itself being encoded in `unicode` If you look closely it's `url` which is created using `_build_body()` is an ASCII string versus `test_url` which has `unicode` characters.

**Comment by tgwizard** [comment]
@tushdante no, `''` is falsy and `'{}'` is truthy. Among strings, only the empty string is falsy.

```python
>>> bool('')
False
>>> bool(json.dumps({}))
True
```

The unicode v str is not the cause of the test failure, as

```python
>>> u'text' == 'text'
True
>>> {u'x': u'y'} == {'x': 'y'}
True
```

The cause of the test failure is an extra `'headers': '{}',` (at least in the second test).

**Comment by tushdante** [comment]
So, it looks like the changes in the `set_replyto()` code are not being picked up by the test. Changing the mock test to 
```
test_url = json.loads('''
            {
                "to[]": ["john@email.com"],
                "toname[]": ["John Doe"],
                "html": "WIN",
                "text": "WIN",
                "subject": "test",
                "files[testFile]": "fileValue",
                "from": "doe@email.com",
                "cc[]": ["cc@email.com"],
                "bcc[]": ["bcc@email.com"],
                "replyto": "John, Doe <john@email.com>"                
            }
            ''')
        test_url['x-smtpapi'] = json.dumps(json.loads('''
            {
                "sub": {
                    "subKey": ["subValue"]
                },
                "section": {
                    "testSection":"sectionValue"
                },
                "category": ["testCategory"],
                "unique_args": {
                    "testUnique":"uniqueValue"
                },
                "filters": {
                    "testFilter": {
                        "settings": {
                            "filter": "filterValue"
                        }
                    }
                },
                "asm_group_id": 42
            }
            '''))
```
However, the `replyto` field shouldn't be populated at all. It's only set in the headers.

**Comment by tushdante** [comment]
Closes #110

**Comment by tushdante** [comment]
Merged

**Comment by to175** [comment]
What is the code to add to have a Replyto please ? I didn't understand there is not this field in the website doc, and is there a second "from" field to set a Name instead of an email ?
thanks :)

**Comment by mbernier** [comment]
@to175 

```
    mail.reply_to = Email("test@example.com")
```
[From the kitchen sink example](https://github.com/sendgrid/sendgrid-python/blob/master/examples/helpers/mail/mail_example.py#L188)

If you have any other questions, please create a new issue - it will cause it to pop to the top of our inboxes :)

**Pull request closed without merging**